epub33/rs/index.html

```diff
@@ -2067,92 +2073,193 @@ <h2>Processing structural semantics</h2>
+								prefix). All valid properties require a name.</li>
+						</ul>
+					</details>
+				</li>
+				<li>
+					<p>Obtain the values of <var>propertyPrefix</var> and <var>propertyName</var> as follows:</p>
+					<ol>
 						<li>
-							<p id="confreq-rs-vocab-property-reference">If the property consists only of a reference,
-								concatenate the prefix URL associated with the <a href="#sec-default-vocab">default
-									vocabulary</a> to the reference.</p>
+							<p>If <var>value</var> does not contain a colon, set <var>propertyName</var> to
+									<var>value</var> and continue processing at the next step.</p>
+							<details class="explanation">
+								<summary>Explanation</summary>
+								<p>If a <code>property</code> value does not have a colon, it does not have a prefix. In
+									this case, the value is drawn from the default vocabulary.</p>
+							</details>
 						</li>
 						<li>
-							<p>
-								<span id="confreq-rs-vocab-property-prefixed">If the property consists of a prefix and
-									reference, concatenate the URL defined for the prefix to the reference.</span>
-								<span id="confreq-rs-vocab-property-invalid-prefix">If the EPUB creator has not defined
-									a matching prefix, and it is not a <a data-cite="epub-33#sec-reserved-prefixes"
-										>reserved prefix</a> [[epub-33]], the property is invalid and reading systems
-									MUST ignore it.</span>
-							</p>
+							<p>Otherwise, if <var>value</var> contains a single colon:</p>
+							<ol>
+								<li>
+									<p>If <var>value</var> begins or ends with a colon, the value is invalid. Return <a
+											data-cite="infra#nulls">null</a>.</p>
+								</li>
+								<li>
+									<p>Otherwise, let <var>tokens</var> be the result of [=strictly split a
+										string|splitting=] <var>value</var> using a colon (U+003A) as the delimiter. Set
+											<var>propertyPrefix</var> to <var>tokens[0]</var> and set
+											<var>propertyName</var> to <var>tokens[1]</var>.</p>
+								</li>
+							</ol>
+							<details class="explanation">
+								<summary>Explanation</summary>
+								<p>If a <code>property</code> value has a single colon, it has a prefix.</p>
+								<p>A colon at the beginning of the value is invalid, however, as it indicates the prefix
+									is not set. Likewise, a colon at the end of the value indicates that there is only a
+									prefix and no value.</p>
+							</details>
 						</li>
 						<li>
-							<p id="confreq-rs-vocab-property-only-prefix">If the property consists only of a prefix
-								(i.e., there is no reference after the colon), the property is invalid and reading
-								systems MUST ignore it.</p>
+							<p>Otherwise, the value is invalid. Return <a data-cite="infra#nulls">null</a>.</p>
+							<details class="explanation">
+								<summary>Explanation</summary>
+								<p>A <code>property</code> value cannot have more than one colon.</p>
```

Comment:
Is that specified somewhere? I thought a colon is an acceptable character in value... In which case the previous step should be modified to say that we split at the first occurrence of a colon...

Reply:
Ya, I over-interpreted the relative path requirement a bit, but it looks to be more complicated than either of us has right now.

According to the [EBNF for the data type](https://w3c.github.io/epub-specs/epub33/core/#sec-property-datatype), the value has to be a [path-relative-scheme-less-URL string](https://url.spec.whatwg.org/#path-relative-scheme-less-url-string).

It says in that definition that the value "does not start with: a [URL-scheme string](https://url.spec.whatwg.org/#url-scheme-string), followed by U+003A (:)".

So if you circle down through the many definitions you get:

- A path-relative-scheme-less URL must be a path-relative-URL string
- A path-relative-URL string must be zero or more URL-path-segment strings, separated from each other by U+002F (/), and not start with U+002F (/)
- A URL-path-segment string must be zero or more URL units excluding U+002F (/) and U+003F (?)

So wouldn't this make a colon illegal in the first URL-path-segment string? If you have one, it has to be after the first slash.

Do we want to try and spec that out, though, or can we leave the validity to the end? The ultimate arbiter is that you get a valid URL string. Multiple colons is a pretty niche case.

Reply:
Maybe we could test this after splitting. So, if there are one or more colons, split on the first. Then before moving to the next step say if `propertyName` contains a colon before a first slash, the value is invalid?

Reply:
If I follow the definition for the property:

- path-relative-scheme-less-URL string -> cannot start with a URL-scheme string followed by a (:) but otherwise is a path-relative-URL string. 
- a URL-scheme is ASCII alpha, followed by ASCII alphanumeric or '+','-','.'
- path-relative-URL string allows for the ':' character.

I.e., the value indeed cannot be "abcd:efgh". But I believe it can be something like "abcd/efg:hij". E.g., I believe this is valid:

```
… prefix="test: http://www.example.org/"
… property="test:abcd/#efg:hij"
```

yielding `http://www.example.org/abcd/#efg:hij`.

This means, I believe, that we should be prepared for the case where the property contains more than one ':' character; meaning that the split must happen at the first occurrence of ':' and then we are done...

Reply:
The latest commit rewrites step 2 to account for multiple colons.

I couldn't use the infra splitting algorithm anymore, as that could lead to multiple chunks which would then have to be recombined. Seemed excessive. I've have it now to split on the first colon and assign the strings before and after to the appropriate variables. I'm assuming splitting is widely enough known that we don't have to cite anything.

Anyway, if there are multiple colons, these changes should only allow valid cases to pass through - where the colon is not in the first segment of the name. I added a couple of examples to the definition to help explain this: "foo:bar/ba:z" is valid while "foo:bar:baz/qux" is not.

If you don't think I've got this right, though, let me know.

Reply:
I am not sure the restriction is right: I believe `foo:bar#ba:z` is also a valid case. Ie, (3) would create a false negative, wouldn't it? There are URI schemes that do not rely on slash character (DID, ISBN, mailto)

I would rather go down the line of

- remove current (3)
- split at the first colon i.e., what is now (4)
- before setting the value of _propertyName_ check whether the value begins with a [URL Scheme string](https://url.spec.whatwg.org/#url-scheme-string) followed by a colon. If true, it is an error and return null, otherwise we set the _propertyName_ and we are done.

Would that work?

Reply:
> check whether the value begins with a [URL Scheme string](https://url.spec.whatwg.org/#url-scheme-string) followed by a colon

Maybe simplest of all would be to say that `propertyName` must be a valid path-relative-scheme-less-URL string, since that's where that exception comes from and will match up with the EBNF. I agree these details of what is valid are better left to the explanation.

Reply:
I've redone step 2 to make a few changes.

- step one no longer says to continue processing at the next step. Even if the value doesn't have a colon, that doesn't mean it's a valid PRSLU string.
- the former third step is gone
- after processing, `propertyReference` is checked that it is not an empty string or an invalid PRSLU

I also renamed `propertyName` to `propertyReference` to match the EBNF naming of the components.

Reply:
Hm, this still may not be right. Looking at the PRSLU definition, an empty string is valid. That means that a prefix-only property value is valid, which also matches the CURIE definition. For example, you could do: `<meta ... scheme="myScheme:"

I know we considered this invalid in the past, as made some weird hacks to create schemes that would have both a prefix and reference (marc:relators comes to mind). But I guess it's valid, so I should take that test out.

Reply:
See also https://github.com/w3c/epub-specs/issues/2382. I've fixed this, too.

Reply:
Yes indeed, I know that in the RDFa/Turtle practice using a CURIE without reference (ie, re-using the prefix as a URL on its own right) is actually fairly frequent. But I thought you have some good reasons to disallow the trailing ':' case...

Otherwise, it looks good to me. Personally, I do not see any reason not to merge this and consider it done...

Reply:
> But I thought you have some good reasons to disallow the trailing ':' case...

Nope... 😄 

The unfortunate truth is that we lacked the requisite expertise to be implementing RDFa in the package document when we tried to build this framework for 3.0. We probably should have abandoned it completely when it was pointed out we were absuing the `about` attribute, but rewriting the whole framework at the last minute likely wouldn't have had much better results.

In this case, I think the definition of an optional prefix followed by a reference is what confused people. Whoever made the first claim that it meant the reference couldn't be an empty string was just accepted as gospel. I know this is the deepest I've ever dug into the fine print, so for that alone this issue has been insightful.

But, yes, I think this now captures everything we need now. Do you think we need a group resolution before merging tihs, since we haven't discussed doing this kind of a rewrite? It's pretty substantive at least in terms of removing normative requirements.

Reply:
I've added a final clarification to the data type definition that a prefix and colon with an empty reference is a valid value. I added this after the statement that an empty string is technically valid but non-conforming, as that's the likely where the confusion arose from.

Reply:
Yes, it would be good to have a WG approval. Except that... @wareid will be out for the next two weeks on vacations, and she should not be worrying about the WG. I do not know when @dauwhe is back, not sure that he will be around next week. Also, there is no real reason these days to hold WG calls, it is all quiet with some people making tests...

Let us wait if @dauwhe is back next week. What we could do is that you send a mail to the WG explaining what has been done, and leave a week to respond/object to the changes. I would hate to leave this PR open for several weeks...

Reply:
>  Also, there is no real reason these days to hold WG calls

Hm, ya, I wonder if I'm overthinking this. We're not adding new requirements, only loosening existing ones. And it's for an optional feature that we're haven't heard from anyone has been implemented.

It's noted in the change log, too, so anyone keeping track of what's going on will find it. Maybe we should just go ahead an merge...

Reply:
I would like to get at least @dauwhe 's reaction (@wareid is out on vacations)